satpy/writers/cf/datasets.py

```diff
@@ -0,0 +1,275 @@
+        Include the original dataset name as a variable attribute in the xr.Dataset.
+    numeric_name_prefix : str, optional
+        Prefix to add the each variable with name starting with a digit.
+        Use '' or None to leave this out.
+
+    Returns
+    -------
+    ds : xr.Dataset
+        A partially CF-compliant xr.Dataset
+    """
+    from satpy.writers.cf.area import (
+        area2cf,
+        assert_xy_unique,
+        has_projection_coords,
+        link_coords,
+        make_alt_coords_unique,
+    )
+    from satpy.writers.cf.dataarray import make_cf_dataarray
+
+    # Create dictionary of input datarrays
+    # --> Since keys=None, it doesn't never retrieve ancillary variables !!!
+    ds_collection = {}
+    for dataarray in list_dataarrays:
+        ds_collection.update(get_extra_ds(dataarray))
+
+    # Check if one DataArray in the collection has 'longitude' or 'latitude'
+    got_lonlats = has_projection_coords(ds_collection)
+
+    # Sort dictionary by keys name
+    ds_collection = dict(sorted(ds_collection.items()))
+
+    dict_dataarrays = {}
+    for dataarray in ds_collection.values():
+        dataarray_type = dataarray.dtype
+        if dataarray_type not in CF_DTYPES:
+            warnings.warn(
+                f'dtype {dataarray_type} not compatible with {CF_VERSION}.',
+                stacklevel=3
+            )
+        # Deep copy the datarray since adding/modifying attributes and coordinates
+        dataarray = dataarray.copy(deep=True)
+
+        # Add CF-compliant area information from the pyresample area
+        # - If include_lonlats=True, add latitude and longitude coordinates
+        # - Add grid_mapping attribute to the DataArray
+        # - Return the CRS DataArray as first list element
+        # - Return the CF-compliant input DataArray as second list element
+        try:
+            list_new_dataarrays = area2cf(dataarray,
+                                          include_lonlats=include_lonlats,
+                                          got_lonlats=got_lonlats)
+        except KeyError:
+            list_new_dataarrays = [dataarray]
+
+        # Ensure each DataArray is CF-compliant
+        # --> NOTE: Here the CRS DataArray is repeatedly overwrited
+        # --> NOTE: If the input list_dataarrays have different pyresample areas with the same name
+        #           area information can be lost here !!!
+        for new_dataarray in list_new_dataarrays:
+            new_dataarray = make_cf_dataarray(new_dataarray,
+                                              epoch=epoch,
+                                              flatten_attrs=flatten_attrs,
+                                              exclude_attrs=exclude_attrs,
+                                              include_orig_name=include_orig_name,
+                                              numeric_name_prefix=numeric_name_prefix)
+            dict_dataarrays[new_dataarray.name] = new_dataarray
+
```

Comment:
The following three calls should be unioned in a single function in my opinion. 
And such function likely refactored.

Reply:
You mean `assert_xy_unique`, `link_coords`, and `make_alt_coords_unique`? If you feel strongly about it then go for it, but I'm not sure they are handling things that are exactly the same so putting them in one function may not make the most sense.